File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/audio/specs/GivenAFileLessThan5Megabytes/WhenStreamingTheFileInOddChunks.java
package com.lasthopesoftware.bluewater.client.library.items.media.audio.specs.GivenAFileLessThan5Megabytes;

import android.net.Uri;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.upstream.DataSpec;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import com.lasthopesoftware.bluewater.client.library.items.media.audio.DiskFileCacheDataSource;
import com.lasthopesoftware.bluewater.client.library.items.media.files.cached.repository.CachedFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.cached.stream.CacheOutputStream;
import com.lasthopesoftware.bluewater.client.library.items.media.files.cached.stream.supplier.ICacheStreamSupplier;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.io.IOException;
import java.util.Random;

import okio.Buffer;
import okio.BufferedSource;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class WhenStreamingTheFileInOddChunks {

	private static final byte[] bytesWritten = new byte[2 * 1024 * 1024];
	private static final byte[] bytes = new byte[2 * 1024 * 1024];

	static {
		new Random().nextBytes(bytes);
	}

	@BeforeClass
	public static void context() throws IOException {
		final ICacheStreamSupplier fakeCacheStreamSupplier = uniqueKey -> new Promise<>(new CacheOutputStream() {
				int numberOfBytesWritten = 0;

				@Override
				public Promise<CacheOutputStream> promiseWrite(byte[] buffer, int offset, int length) {
					return new Promise<>(this);
				}

				@Override
				public Promise<CacheOutputStream> promiseTransfer(BufferedSource bufferedSource) {
					try {
						while (numberOfBytesWritten < bytesWritten.length) {
							int read = bufferedSource.read(bytesWritten, numberOfBytesWritten, bytesWritten.length - numberOfBytesWritten);
							if (read == -1) return new Promise<>(this);
							numberOfBytesWritten += read;
						}
					} catch (IOException e) {
						e.printStackTrace();
					}
					return new Promise<>(this);
				}

				@Override
				public Promise<CachedFile> commitToCache() {
					return new Promise<>(new CachedFile());
				}

				@Override
				public Promise<CacheOutputStream> flush() {
					return new Promise<>(this);
				}

				@Override
				public void close() throws IOException {

				}
			});

		final Buffer buffer = new Buffer();
		buffer.write(bytes);

		final HttpDataSource dataSource = mock(HttpDataSource.class);
		when(dataSource.read(any(), anyInt(), anyInt()))
			.then(invocation -> {
				if (buffer.exhausted()) return C.RESULT_END_OF_INPUT;

				final byte[] bytes = invocation.getArgument(0);
				final int offset = invocation.getArgument(1);
				final int length = invocation.getArgument(2);
				return buffer.read(bytes, offset, length);
			});

		final DiskFileCacheDataSource diskFileCacheDataSource =
			new DiskFileCacheDataSource(
				dataSource,
				fakeCacheStreamSupplier);

		diskFileCacheDataSource.open(new DataSpec(Uri.parse("http://my-server/file"), 0, 2 * 1024 * 1024, "1"));

		final Random random = new Random();
		int readResult;
		do {
			final byte[] bytes = new byte[random.nextInt(1000000)];
			readResult = diskFileCacheDataSource.read(bytes, 0, bytes.length);
		} while (readResult != C.RESULT_END_OF_INPUT);
	}

	@Test
	public void thenTheEntireFileIsWritten() {
		Assert.assertArrayEquals(bytes, bytesWritten);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/caching/GivenAFileLessThan1Megabyte/WhenStreamingTheFileInOddChunks.kt
package com.lasthopesoftware.bluewater.client.playback.caching.GivenAFileLessThan1Megabyte

import android.net.Uri
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.upstream.DataSpec
import com.google.android.exoplayer2.upstream.HttpDataSource
import com.lasthopesoftware.bluewater.client.browsing.files.cached.repository.CachedFile
import com.lasthopesoftware.bluewater.client.browsing.files.cached.stream.CacheOutputStream
import com.lasthopesoftware.bluewater.client.browsing.files.cached.stream.supplier.SupplyCacheStreams
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.playback.caching.datasource.EntireFileCachedDataSource
import com.lasthopesoftware.bluewater.shared.promises.extensions.DeferredPromise
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import okio.Buffer
import okio.BufferedSource
import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.assertArrayEquals
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.*

@RunWith(RobolectricTestRunner::class)
class WhenStreamingTheFileInOddChunks {
	companion object {
		private const val libraryId = 393
		private var writtenLibraryId: LibraryId? = null
		private val bytesWritten = ByteArray(512 * 1024)
		private val bytes by lazy { ByteArray(512 * 1024).also { Random().nextBytes(it) } }
		private var committedToCache = false

		@BeforeClass
		@JvmStatic
		fun context() {
			val deferredCommit = DeferredPromise(CachedFile() as CachedFile?)

			val fakeCacheStreamSupplier =
				object : SupplyCacheStreams {
					override fun promiseCachedFileOutputStream(libraryId: LibraryId, uniqueKey: String): Promise<CacheOutputStream> {
						writtenLibraryId = libraryId
						return Promise<CacheOutputStream>(object : CacheOutputStream {
							var numberOfBytesWritten = 0
							override fun promiseWrite(
								buffer: ByteArray,
								offset: Int,
								length: Int
							): Promise<CacheOutputStream> =
								Promise<CacheOutputStream>(this)

							override fun promiseTransfer(bufferedSource: BufferedSource): Promise<CacheOutputStream> {
								while (numberOfBytesWritten < bytesWritten.size) {
									val read = bufferedSource.read(
										bytesWritten,
										numberOfBytesWritten,
										bytesWritten.size - numberOfBytesWritten
									)
									if (read == -1) return Promise<CacheOutputStream>(this)
									numberOfBytesWritten += read
								}
								return Promise<CacheOutputStream>(this)
							}

							override fun commitToCache(): Promise<CachedFile?> {
								committedToCache = true
								deferredCommit.resolve()
								return deferredCommit
							}

							override fun flush(): Promise<CacheOutputStream> {
								return Promise<CacheOutputStream>(this)
							}

							override fun close() {}
						})
					}
				}
			val buffer = Buffer()
			buffer.write(bytes)
			val dataSource = mockk<HttpDataSource>(relaxUnitFun = true).apply {
				every { open(any()) } returns bytes.size.toLong()

				every { read(any(), any(), any()) } answers {
					val bytesToRead = arg<Int>(2)
					val offset = arg<Int>(1)
					var bytesRead = 0
					while (bytesRead < bytesToRead) {
						val bufferRead = buffer.read(arg(0), offset + bytesRead, bytesToRead - bytesRead)
						if (bufferRead == -1) {
							if (bytesRead == 0) bytesRead = -1
							break
						}

						bytesRead += bufferRead
					}
					bytesRead
				}
			}
			val diskFileCacheDataSource = EntireFileCachedDataSource(
				LibraryId(libraryId),
				dataSource,
				fakeCacheStreamSupplier,
			)
			diskFileCacheDataSource.open(
				DataSpec.Builder()
					.setUri(Uri.parse("http://my-server/file"))
					.setPosition(0)
					.setLength(C.LENGTH_UNSET.toLong()).setKey("1")
					.build()
			)
			val random = Random()
			var readResult: Int
			do {
				val bytes = ByteArray(random.nextInt(1000000))
				readResult = diskFileCacheDataSource.read(bytes, 0, bytes.size)
			} while (readResult != C.RESULT_END_OF_INPUT)
			diskFileCacheDataSource.close()

			deferredCommit.toExpiringFuture().get()
		}
	}

	@Test
	fun `then the correct library is written to`() {
		assertThat(writtenLibraryId?.id).isEqualTo(libraryId)
	}

	@Test
	fun thenTheEntireFileIsWritten() {
		assertArrayEquals(bytes, bytesWritten)
	}

	@Test
	fun thenTheFileIsCached() {
		assertThat(committedToCache).isTrue
	}
}
